File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.java
import java.util.*;

public class Main {
    public static int longestIncreasingSubsequence(int[] arr) {
        List<Integer> lis = new ArrayList<>();
        
        for (int num : arr) {
            // 이진 탐색을 이용하여 lis에서 적절한 위치를 찾는다.
            int pos = Collections.binarySearch(lis, num);
            
            // binarySearch는 값이 없으면 음수 위치를 반환하므로 이를 처리
            if (pos < 0) pos = -(pos + 1);
            
            // 위치가 이미 있으면 값을 바꿔준다.
            if (pos < lis.size()) {
                lis.set(pos, num);
            } else {
                // 새롭게 추가한다.
                lis.add(num);
            }
        }
        
        // lis의 크기가 가장 긴 증가하는 부분 수열의 길이
        return lis.size();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        // 입력 받기
        int n = sc.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }
        
        // 결과 출력
        System.out.println(longestIncreasingSubsequence(arr));
    }
}

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.py
import bisect

def longest_increasing_subsequence(arr):
    # LIS를 구성하는 부분 수열을 저장하는 리스트
    lis = []
    
    for num in arr:
        # num이 lis에 들어갈 위치를 찾는다.
        pos = bisect.bisect_left(lis, num)
        
        # lis에 num이 들어갈 위치가 존재하면 해당 위치의 값을 num으로 대체
        if pos < len(lis):
            lis[pos] = num
        else:
            # lis에 추가
            lis.append(num)
    
    # lis의 길이가 가장 긴 증가하는 부분 수열의 길이
    return len(lis)

# 입력
n = int(input())  # 카드의 개수
arr = list(map(int, input().split()))  # 수열

# 결과 출력
print(longest_increasing_subsequence(arr))
